### PR TITLE
SAK-41505 Added direct memberships to GB sections columns

### PR DESCRIPTION
**Pull request by adrianfish**
https://jira.sakaiproject.org/browse/SAK-41505

### CODE REVIEW

**Comment by ern** [comment]
Is this a bug in the SectionManager service?

**Comment by ottenhoff** [comment]
I don't think it's a bug in SectionManager service... just multiple ways a student can be joined to a section via CMAPI ....

**Comment by adrianfish** [comment]
Like Sam says. It's like there's a concept of enrollments in a top level course section, but then the student participation in the child 'sections' are just termed 'memberships'. I'll be honest and say I don't fully understand the way the CM api is supposed to work.

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java [hunk 1]
```diff
@@ -1033,18 +1046,59 @@ public GradeSaveResponse saveExcuse(final Long assignmentId, final String studen
...
-				String userId = er.getUser().getUserUid();
+		final Map<String, List<String>> userSections = new HashMap<>();
+
+		String[] sectionIds = null;
+
+		try {
+			sectionIds = groupProvider.unpackId(siteService.getSite(siteId).getProviderGroupId());
+		} catch (IdUnusedException idue) {}
+
+		if (sectionIds == null || sectionIds.length == 0) {
```
Comment (bjones86) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java]: You can use `ArrayUtils.isEmpty()` here

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java [hunk 2]
```diff
@@ -1033,18 +1046,59 @@ public GradeSaveResponse saveExcuse(final Long assignmentId, final String studen
...
+			if (section == null) {
+				log.debug("Section '{}'  not found, skipping ...", sectionId);
+				continue;
+			}
+
+			EnrollmentSet enrollmentSet = section.getEnrollmentSet();
+
+			Set<Membership> memberships = courseManagementService.getSectionMemberships(sectionId);
+
+			if ((memberships == null || memberships.size() == 0) && enrollmentSet == null) {
```
Comment (bjones86) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java]: `CollectionUtils.isEmpty()`

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java (hunk 1 above)
Comment (ottenhoff) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java]: Implemented

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java (hunk 2 above)
Comment (ottenhoff) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java]: Implemented

**Pull request merged**